packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -321,6 +331,137 @@ export class Wallet extends EventEmitter<WalletEvents> {
     });
   }
 
+  private emitObjectiveEvents(result: MultipleChannelOutput | SingleChannelOutput): void {
+    if (isMultipleChannelOutput(result)) {
+      // Receiving messages from other participants may have resulted in completed objectives
+      for (const o of result.completedObjectives) {
+        this.emit('ObjectiveCompleted', o);
+      }
+
+      // Receiving messages from other participants may have resulted in new proposed objectives
+      for (const o of result.newObjectives) {
+        if (o.status === 'pending') {
+          this.emit('ObjectiveProposed', o);
+        }
+      }
+    } else {
+      if (hasNewObjective(result)) {
+        if (result.newObjective.status === 'pending') {
+          this.emit('ObjectiveProposed', result.newObjective);
+        }
+      }
+    }
+  }
+
+  /**
+   * Waits for the transactions.wait() promise to resolve on all transaction responses.
+   * @param response
+   */
+  private async waitForTransactions(
+    response: Promise<providers.TransactionResponse[]>
+  ): Promise<void> {
+    const transactions = await response;
+    await Promise.all(transactions.map(tr => tr.wait()));
+  }
+
+  /**
+   * Emits events, sends messages and requests transactions based on the output of the engine.
+   * @param output
+   */
+  private async handleEngineOutput(
+    output: MultipleChannelOutput | SingleChannelOutput
+  ): Promise<void> {
+    this.emitObjectiveEvents(output);
+    await this._messageService.send(getMessages(output));
+    await this.waitForTransactions(this._chainService.handleChainRequests(output.chainRequests));
+  }
+
+  /**
+   * Creates a listener that will be subscribed to the chain service.
+   * It creates a listener for a specific channel id.
+   * @param channelIdToListenFor The channel Id the listener is for
+   * @returns A chain event subscriber that can be passed into the chain service.
+   */
+  private createChainListener(channelIdToListenFor: string): ChainEventSubscriberInterface {
```

Comment:
this now creates a new chain listener for each channel. why's that?

it seems like that choice will cause the chain service to call `holdingUpdated` a quadratic number of times, as a function of the number of channels?

Reply:
I think you're right. We need to pass in a listener every time we call `registerChannel` but we can re-use the same listener and just check the channel id

Reply:
Updated in 19a3e2dbedcd9153b142f49d3f208ce009e6ceb1